File: internal/export/export.go
// Package export renders a stored run. Cue segmentation happens here because
// the recognition API does not segment: a multi-sentence take comes back as one
// sentence spanning the whole file, both non-streaming and over SSE. The only
// segmentation signal available is per-word punctuation and inter-word gaps.
package export

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/celados/vox/internal/dashscope"
	"github.com/celados/vox/internal/run"
)

// Cue conventions. Fixed rather than exposed as flags: these are subtitle
// typesetting norms, not per-call decisions.
const (
	maxCueRunes    = 28
	maxCueDuration = 6 * time.Second
	// gapSplit is the inter-word silence that ends a cue even mid-clause.
	gapSplit = 400 * time.Millisecond
)

// terminators end a cue outright; separators end one only if it is long enough.
const (
	terminators = "。！？!?…"
	separators  = "，、；：,;:"
)

type Cue struct {
	Index int
	Begin time.Duration
	End   time.Duration
	Text  string
}

// Segment groups words into subtitle cues.
func Segment(words []dashscope.ASRWord) []Cue {
	var cues []Cue
	var buf strings.Builder
	var begin, end time.Duration
	var runeCount int
	started := false

	flush := func() {
		text := strings.TrimSpace(buf.String())
		if text == "" {
			buf.Reset()
			started = false
			runeCount = 0
			return
		}
		cues = append(cues, Cue{Index: len(cues) + 1, Begin: begin, End: end, Text: text})
		buf.Reset()
		started = false
		runeCount = 0
	}

	for i, w := range words {
		wordBegin := time.Duration(w.BeginTime) * time.Millisecond
		wordEnd := time.Duration(w.EndTime) * time.Millisecond

		// A long silence before this word closes the pending cue.
		if started && wordBegin-end >= gapSplit {
			flush()
		}
		if !started {
			begin, started = wordBegin, true
		}
		buf.WriteString(w.Text)
		buf.WriteString(w.Punctuation)
		runeCount += len([]rune(w.Text))
		end = wordEnd

		last := i == len(words)-1
		switch {
		case last:
			flush()
		case strings.ContainsAny(w.Punctuation, terminators):
			flush()
		case strings.ContainsAny(w.Punctuation, separators) && runeCount >= maxCueRunes/2:
			flush()
		case runeCount >= maxCueRunes:
			flush()
		case end-begin >= maxCueDuration:
			flush()
		}
	}
	if started {
		flush()
	}
	return cues
}

// Render produces the requested format. Formats are the whole contract; adding
// one here is a surface change.
func Render(rec *run.Record, format string) (string, error) {
	switch format {
	case "txt":
		return rec.Result.Text + "\n", nil
	case "json":
		data, err := json.MarshalIndent(rec, "", "  ")
		return string(data) + "\n", err
	case "md":
		return renderMarkdown(rec), nil
	case "srt":
		return renderSRT(Segment(rec.Result.Words)), nil
	case "vtt":
		return renderVTT(Segment(rec.Result.Words)), nil
	default:
		return "", fmt.Errorf("unknown format %q", format)
	}
}

func renderMarkdown(rec *run.Record) string {
	var b strings.Builder
	fmt.Fprintf(&b, "---\ntype: Transcript\nsid: %s\nsource: %s\nmodel: %s\n", rec.Meta.SID, rec.Meta.Source, rec.Meta.Model)
	if rec.Meta.Vocab != "" {
		fmt.Fprintf(&b, "vocab: %s\n", rec.Meta.Vocab)
	}
	fmt.Fprintf(&b, "duration: %s\ngenerated: { by: vox, at: %s }\n---\n\n",
		formatDuration(time.Duration(rec.Meta.Size.Duration)*time.Second),
		rec.Meta.Created.UTC().Format(time.RFC3339))

	// Paragraphs follow the same cue segmentation, merged until a sentence ends,
	// so prose stays readable instead of one wall of text.
	var para strings.Builder
	for _, cue := range Segment(rec.Result.Words) {
		para.WriteString(cue.Text)
		if strings.ContainsAny(lastRune(cue.Text), terminators) {
			b.WriteString(para.String())
			b.WriteString("\n\n")
			para.Reset()
		}
	}
	if para.Len() > 0 {
		b.WriteString(para.String())
		b.WriteString("\n")
	}
	return b.String()
}

func renderSRT(cues []Cue) string {
	var b strings.Builder
	for _, cue := range cues {
		fmt.Fprintf(&b, "%d\n%s --> %s\n%s\n\n",
			cue.Index, srtTime(cue.Begin), srtTime(cue.End), cue.Text)
	}
	return b.String()
}

func renderVTT(cues []Cue) string {
	var b strings.Builder
	b.WriteString("WEBVTT\n\n")
	for _, cue := range cues {
		fmt.Fprintf(&b, "%s --> %s\n%s\n\n",
			vttTime(cue.Begin), vttTime(cue.End), cue.Text)
	}
	return b.String()
}

func srtTime(d time.Duration) string {
	h, m, s, ms := splitDuration(d)
	return fmt.Sprintf("%02d:%02d:%02d,%03d", h, m, s, ms)
}

func vttTime(d time.Duration) string {
	h, m, s, ms := splitDuration(d)
	return fmt.Sprintf("%02d:%02d:%02d.%03d", h, m, s, ms)
}

func splitDuration(d time.Duration) (h, m, s, ms int) {
	total := int(d.Milliseconds())
	return total / 3600000, total / 60000 % 60, total / 1000 % 60, total % 1000
}

func formatDuration(d time.Duration) string {
	h, m, s, _ := splitDuration(d)
	if h > 0 {
		return fmt.Sprintf("%02d:%02d:%02d", h, m, s)
	}
	return fmt.Sprintf("%02d:%02d", m, s)
}

func lastRune(s string) string {
	r := []rune(s)
	if len(r) == 0 {
		return ""
	}
	return string(r[len(r)-1])
}

File: src/export.ts
import { stringify } from "yaml";

import type { AsrWord, Sentence } from "./dashscope.ts";
import type { RunRecord } from "./run-store.ts";

export const MAX_CUE_RUNES = 28;
export const MAX_CUE_DURATION_MS = 6_000;
export const GAP_SPLIT_MS = 400;
export const MIN_WORD_MS = 40;

const TERMINATORS = "。！？!?…";
const SEPARATORS = "，、；：,;:";

export type Cue = {
  index: number;
  beginMs: number;
  endMs: number;
  text: string;
};

export function segment(words: AsrWord[]): Cue[] {
  const sanitized = sanitize(words);
  const cues: Cue[] = [];
  let buf = "";
  let begin = 0;
  let end = 0;
  let runeCount = 0;
  let started = false;

  const flush = () => {
    const text = buf.trim();
    buf = "";
    runeCount = 0;
    started = false;
    if (!text) return;
    cues.push({ index: cues.length + 1, beginMs: begin, endMs: end, text });
  };

  for (let i = 0; i < sanitized.length; i++) {
    const word = sanitized[i]!;
    const wordBegin = word.begin_time;
    const wordEnd = word.end_time;
    if (started && wordBegin - end >= GAP_SPLIT_MS) flush();
    if (
      started &&
      (runeCount + [...word.text].length > MAX_CUE_RUNES || wordEnd - begin > MAX_CUE_DURATION_MS)
    ) {
      flush();
    }
    if (!started) {
      begin = wordBegin;
      started = true;
    }
    buf += word.text + word.punctuation;
    runeCount += [...word.text].length;
    end = wordEnd;

    const last = i === sanitized.length - 1;
    if (last) flush();
    else if (containsAny(word.punctuation, TERMINATORS)) flush();
    else if (containsAny(word.punctuation, SEPARATORS) && runeCount >= MAX_CUE_RUNES / 2) flush();
    else if (runeCount >= MAX_CUE_RUNES) flush();
    else if (end - begin >= MAX_CUE_DURATION_MS) flush();
  }
  if (started) flush();
  return cues;
}

function sanitize(words: AsrWord[]): AsrWord[] {
  const out: AsrWord[] = [];
  let prevEnd = 0;
  for (const word of words) {
    if (!word.text.trim() && !word.punctuation) continue;
    const next = { ...word };
    if (next.begin_time < prevEnd) next.begin_time = prevEnd;
    if (next.end_time <= next.begin_time) next.end_time = next.begin_time + MIN_WORD_MS;
    prevEnd = next.end_time;
    out.push(next);
  }
  return out;
}

export function render(rec: RunRecord, format: string): string {
  switch (format) {
    case "txt":
      return rec.result.text + "\n";
    case "json":
      return JSON.stringify(rec, null, 2) + "\n";
    case "md":
      return renderMarkdown(rec);
    case "srt":
      return renderSrt(cuesOf(rec));
    case "vtt":
      return renderVtt(cuesOf(rec));
    default:
      throw new Error(`unknown format "${format}"`);
  }
}

export function cuesOf(rec: RunRecord): Cue[] {
  if (rec.result.sentences && rec.result.sentences.length > 0) {
    return fromSentences(rec.result.sentences);
  }
  if (!rec.result.text.trim()) return [];
  return [
    {
      index: 1,
      beginMs: 0,
      endMs: rec.meta.size.duration * 1000,
      text: rec.result.text,
    },
  ];
}

function fromSentences(sentences: Sentence[]): Cue[] {
  const cues: Cue[] = [];
  for (const sentence of sentences) {
    let text = sentence.text.trim();
    if (!text) continue;
    if (sentence.speaker) text = `[${sentence.speaker}] ${text}`;
    const duration = sentence.end_time - sentence.begin_time;
    if ([...text].length <= MAX_CUE_RUNES && duration <= MAX_CUE_DURATION_MS) {
      cues.push({
        index: cues.length + 1,
        beginMs: sentence.begin_time,
        endMs: sentence.end_time,
        text,
      });
      continue;
    }
    let first = true;
    for (const cue of segment(sentence.words ?? [])) {
      const next = { ...cue, index: cues.length + 1 };
      // Prefix the first split cue of this sentence, not of the whole file.
      if (sentence.speaker && first) {
        next.text = `[${sentence.speaker}] ${next.text}`;
        first = false;
      }
      cues.push(next);
    }
  }
  return cues;
}

export function renderMarkdown(rec: RunRecord): string {
  const head = stringify({
    type: "Transcript",
    sid: rec.meta.sid,
    source: rec.meta.source,
    model: rec.meta.model,
    ...(rec.meta.vocab ? { vocab: rec.meta.vocab } : {}),
    duration: formatDuration(rec.meta.size.duration * 1000),
    generated: { by: "vox", at: toRfc3339(rec.meta.created) },
  });

  let body = "";
  const units = cuesOf(rec);
  if (units.length === 0) {
    body = rec.result.text.trim() + "\n";
    return `---\n${head}---\n\n${body}`;
  }

  let para = "";
  for (const cue of units) {
    para += cue.text;
    if (containsAny(lastRune(cue.text), TERMINATORS)) {
      body += para + "\n\n";
      para = "";
    }
  }
  if (para) body += para + "\n";
  return `---\n${head}---\n\n${body}`;
}

export function renderSrt(cues: Cue[]): string {
  let out = "";
  for (const cue of cues) {
    out += `${cue.index}\n${srtTime(cue.beginMs)} --> ${srtTime(cue.endMs)}\n${cue.text}\n\n`;
  }
  return out;
}

export function renderVtt(cues: Cue[]): string {
  let out = "WEBVTT\n\n";
  for (const cue of cues) {
    out += `${vttTime(cue.beginMs)} --> ${vttTime(cue.endMs)}\n${cue.text}\n\n`;
  }
  return out;
}

export function srtTime(ms: number): string {
  const parts = splitMs(ms);
  return `${pad(parts.h, 2)}:${pad(parts.m, 2)}:${pad(parts.s, 2)},${pad(parts.ms, 3)}`;
}

export function vttTime(ms: number): string {
  const parts = splitMs(ms);
  return `${pad(parts.h, 2)}:${pad(parts.m, 2)}:${pad(parts.s, 2)}.${pad(parts.ms, 3)}`;
}

function formatDuration(ms: number): string {
  const parts = splitMs(ms);
  if (parts.h > 0) return `${pad(parts.h, 2)}:${pad(parts.m, 2)}:${pad(parts.s, 2)}`;
  return `${pad(parts.m, 2)}:${pad(parts.s, 2)}`;
}

function splitMs(ms: number): { h: number; m: number; s: number; ms: number } {
  const total = Math.trunc(ms);
  return {
    h: Math.floor(total / 3_600_000),
    m: Math.floor(total / 60_000) % 60,
    s: Math.floor(total / 1000) % 60,
    ms: total % 1000,
  };
}

function pad(n: number, width: number): string {
  return String(n).padStart(width, "0");
}

function lastRune(text: string): string {
  const chars = [...text];
  return chars[chars.length - 1] ?? "";
}

function containsAny(text: string, chars: string): boolean {
  for (const ch of text) {
    if (chars.includes(ch)) return true;
  }
  return false;
}

function toRfc3339(created: string): string {
  const date = new Date(created);
  if (Number.isNaN(date.getTime())) return created;
  return date.toISOString();
}
